src/Workspaces/Core/Portable/Shared/Extensions/ISymbolExtensions.cs

```diff
@@ -645,38 +645,36 @@ private static bool ElementNameIs(XElement element, string name)
     public static ImmutableArray<T> FilterToVisibleAndBrowsableSymbols<T>(
         this ImmutableArray<T> symbols, bool hideAdvancedMembers, Compilation compilation) where T : ISymbol
     {
-        symbols = symbols.RemoveOverriddenSymbolsWithinSet();
+        var overriddenSymbols = MetadataUnifyingSymbolHashSet.AllocateFromPool();
+
+        foreach (var symbol in symbols)
+        {
+            var overriddenMember = symbol.GetOverriddenMember();
+            if (overriddenMember != null)
+                overriddenSymbols.Add(overriddenMember);
+        }
 
         // Since all symbols are from the same compilation, find the required attribute
         // constructors once and reuse.
         var editorBrowsableInfo = new EditorBrowsableInfo(compilation);
 
         // PERF: HasUnsupportedMetadata may require recreating the syntax tree to get the base class, so first
         // check to see if we're referencing a symbol defined in source.
-        return symbols.WhereAsArray((s, arg) =>
+        var filteredSymbols = symbols.WhereAsArray(static (s, arg) =>
             // Check if symbol is namespace (which is always visible) first to avoid realizing all locations
             // of each namespace symbol, which might end up allocating in LOH
+            !arg.overriddenSymbols.Contains(s) &&
             (s is INamespaceSymbol || s.Locations.Any(static loc => loc.IsInSource) || !s.HasUnsupportedMetadata) &&
             !s.IsDestructor() &&
             s.IsEditorBrowsable(
                 arg.hideAdvancedMembers,
                 arg.editorBrowsableInfo.Compilation,
                 arg.editorBrowsableInfo),
-            (hideAdvancedMembers, editorBrowsableInfo));
-    }
+            arg: (hideAdvancedMembers, editorBrowsableInfo, overriddenSymbols));
 
-    private static ImmutableArray<T> RemoveOverriddenSymbolsWithinSet<T>(this ImmutableArray<T> symbols) where T : ISymbol
-    {
-        var overriddenSymbols = new MetadataUnifyingSymbolHashSet();
-
-        foreach (var symbol in symbols)
-        {
-            var overriddenMember = symbol.GetOverriddenMember();
-            if (overriddenMember != null && !overriddenSymbols.Contains(overriddenMember))
-                overriddenSymbols.Add(overriddenMember);
-        }
+        MetadataUnifyingSymbolHashSet.ClearAndFree(overriddenSymbols);
```

Comment:
try/finaly?  or perhaps a disposable versino of this?

Reply:
Sure!